### ISSUE TITLE
Detecting YCCK and CMYK color space

### ISSUE THREAD

**Issue by aerpenbeck** [haraldk/TwelveMonkeys#571]
I'm having trouble how to correctly determine the color space for YCCK and CMYK images.

Examples are [cmyk-sample-multiple-chunk-icc.jpg](https://github.com/haraldk/TwelveMonkeys/blob/master/imageio/imageio-jpeg/src/test/resources/jpeg/cmyk-sample-multiple-chunk-icc.jpg) and [cmyk-sample-custom-icc-bright.jpg](https://github.com/haraldk/TwelveMonkeys/blob/master/imageio/imageio-jpeg/src/test/resources/jpeg/cmyk-sample-custom-icc-bright.jpg) from TwelveMonkeys own test images.
Another example is [this one](https://drive.google.com/file/d/14LyilhYAGa0S4js09tMD_B0cx02m7daW/view?usp=sharing).

When I read the color space from the image metadata (using `ImageReader#getImageMetadata()` and the `Chroma/ColorSpaceType` node) I get `YCCK`.
This information seems to be read from the Adobe JPEG directory (Color Transform).
But color space from ICC profile for these images is `CMYK`.

ImageMagick 6.9.7-4 reports `Colorspace: CMYK`.
exiftool reports `Color Mode: CMYK` and `Color Transform: YCCK`

What would be the correct approach to take when I want to inform the user of my app about the color space of a given image?

I was reading the explanation to [issue 545](https://github.com/haraldk/TwelveMonkeys/issues/545) but still don't quite understand why the implementation in `com.twelvemonkeys.imageio.plugins.jpeg.JPEGImageReader#getImageTypes` supports both YCCK and CMYK while the switch in `com.twelvemonkeys.imageio.plugins.jpeg.JPEGImageReader#getRawImageType` only handles CMYK (and therefore returns `null`).

Thanks for helping to understand this.

**Comment by haraldk** [comment]
Hi Arno,

For the question:

> What would be the correct approach to take when I want to inform the user of my app about the color space of a given image?

The short answer is; For the user it's CMYK. The color space is really CMYK regardless of the encoding. In the same way "normal" JPEGs are RGB, even though they are encoded as YCbCr. 

ICC profiles embedded in JPEGs will always be for RGB or CMYK even though the image data may be encoded as YCbCr or YCCK. 

Now, the reasons for the design/implementation choices for `ColorSpaceType` node in the metadata, and the `getRawImageType/getImageTypes` methods, is a slightly longer answer. 

First, keep in mind that it's possible to decode images as `BufferedImage`s and `Raster`s. The `readRaster` method of `ImageReader` is documented: 

> Returns a new `Raster` object containing the raw pixel data from the image stream, without any color conversion applied. The application must determine how to interpret the pixel data by other means.

The `com.sun....JPEGImageReader` has always returned the YCbCr values for YCbCr-encoded JPEGs, so I kept this behavior for YCCK JPEGs. And as the `com.sun..`reader reported the encoded color space in the metadata as the "other means" for applications to determine this encoding, I kept that as well. 

For the `getRawImageType`, the implementation for `com.sun...JPEGImageReader` and the `ImageReader` API doc was actually changed (I think for Java 9), from: 

> For example, for a JPEG image the raw type might have a YCbCr color space even though the image would conventionally be transformed into an RGB color space prior to display. 

(but as Java has no built-in YCbCr `ColorSpace` implementation, the method would return `null` for such cases). 

To: 

> If there is no close match then a type which preserves the most information from the image should be returned.

...and the implementation for `JPEGImageReader` was changed to return an `ImageType` using normal sRGB color space.

See [JDK-8143562](https://bugs.openjdk.java.net/browse/JDK-8143562) for discussion and details.

In either case, the only way for client code to properly distinguish between YCC encoded and "plain" color spaces, is the metadata.


Hope that helps! 


-- 
Harald K
